File: .NET/Examples/WindowsDesktop/Components/InterpreterModeControl.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using UnderAutomation.UniversalRobots;
using UnderAutomation.UniversalRobots.Common;
using UnderAutomation.UniversalRobots.InterpreterMode;

public partial class InterpreterModeControl : UserControl, IUserControl
{
    UR _ur;
    public InterpreterModeControl(UR ur)
    {
        _ur = ur;

        InitializeComponent();
    }

    #region IUserControl
    public string Title => "Interpreter mode";

    public bool FeatureEnabled => _ur.InterpreterMode.Connected;

    public void PeriodicUpdate()
    {
        pnlMain.Enabled = FeatureEnabled;
    }

    public void OnClose() { }

    public void OnOpen() { }
    #endregion

    // Append command response to the console text area
    private void Log(CommandResponse response)
    {
        txtConsole.AppendText(response.Command);
        if (response.Status == CommandResponseStatus.Error)
        {
            txtConsole.SelectionColor = Color.Red;
        }
        else if (response.Status == CommandResponseStatus.Discard)
        {
            txtConsole.SelectionColor = Color.Orange;
        }
        else
        {
            txtConsole.SelectionColor = Color.Green;
        }
        
        txtConsole.AppendText("\r\n  Status: ");
        txtConsole.AppendText(response.Status.ToString());
        if (response.Status != CommandResponseStatus.Error)
        {
            txtConsole.AppendText("\r\n  Id: ");
            txtConsole.AppendText(response.Id.ToString());
        }
        txtConsole.AppendText("\r\n  Body: ");
        txtConsole.AppendText(response.Body);

        txtConsole.SelectionColor = Color.White;
        txtConsole.AppendText("\r\n\r\n> ");
        txtConsole.ScrollToCaret();
    }


    private void btnSendCommand_Click(object sender, EventArgs e)
    {
        var response = _ur.InterpreterMode.ExecuteCommand(txtCommand.Text);
        Log(response);
    }

    private void btnEndInterpreter_Click(object sender, EventArgs e)
    {
        var response = _ur.InterpreterMode.EndInterpreter();
        Log(response);
    }

    private void btnClearInterpreter_Click(object sender, EventArgs e)
    {
        var response = _ur.InterpreterMode.ClearInterpreter();
        Log(response);
    }

    private void btnAbort_Click(object sender, EventArgs e)
    {
        var response = _ur.InterpreterMode.Abort();
        Log(response);
    }

    private void btnSkipBuffer_Click(object sender, EventArgs e)
    {
        var response = _ur.InterpreterMode.SkipBuffer();
        Log(response);
    }

    private void btnLastExecuted_Click(object sender, EventArgs e)
    {
        var response = _ur.InterpreterMode.StateLastExecuted();
        Log(response);
    }

    private void btnLastInterpreted_Click(object sender, EventArgs e)
    {
        var response = _ur.InterpreterMode.StateLastInterpreted();
        Log(response);
    }

    private void btnLastCleared_Click(object sender, EventArgs e)
    {
        var response = _ur.InterpreterMode.StateLastCleared();
        Log(response);
    }

    private void btnLastUnexecuted_Click(object sender, EventArgs e)
    {
        var response = _ur.InterpreterMode.StateLastUnexecuted();
        Log(response);
    }
}

File: UnderAutomation.UniversalRobots.Showcase.Forms.vb/Components/InterpreterModeControl.vb
Imports System
Imports System.Drawing
Imports System.Windows.Forms
Imports UnderAutomation.UniversalRobots
Imports UnderAutomation.UniversalRobots.InterpreterMode

Public Partial Class InterpreterModeControl
    Inherits UserControl
    Implements IUserControl
    Private _ur As UR
    Public Sub New(ur As UR)
        _ur = ur

        InitializeComponent()
    End Sub

#Region "IUserControl"
    Public ReadOnly Property Title As String Implements IUserControl.Title
        Get
            Return "Interpreter mode"
        End Get
    End Property

    Public ReadOnly Property FeatureEnabled As Boolean Implements IUserControl.FeatureEnabled
        Get
            Return _ur.InterpreterMode.Connected
        End Get
    End Property

    Public Sub PeriodicUpdate() Implements IUserControl.PeriodicUpdate
        pnlMain.Enabled = FeatureEnabled
    End Sub

    Public Sub OnClose() Implements IUserControl.OnClose
    End Sub

    Public Sub OnOpen() Implements IUserControl.OnOpen
    End Sub
#End Region

    ' Append command response to the console text area
    Private Sub Log(response As CommandResponse)
        txtConsole.AppendText(response.Command)
        If response.Status = CommandResponseStatus.Error Then
            txtConsole.SelectionColor = Color.Red
        ElseIf response.Status = CommandResponseStatus.Discard Then
            txtConsole.SelectionColor = Color.Orange
        Else
            txtConsole.SelectionColor = Color.Green
        End If

        txtConsole.AppendText(Microsoft.VisualBasic.Constants.vbCrLf & "  Status: ")
        txtConsole.AppendText(response.Status.ToString())
        If response.Status <> CommandResponseStatus.Error Then
            txtConsole.AppendText(Microsoft.VisualBasic.Constants.vbCrLf & "  Id: ")
            txtConsole.AppendText(response.Id.ToString())
        End If
        txtConsole.AppendText(Microsoft.VisualBasic.Constants.vbCrLf & "  Body: ")
        txtConsole.AppendText(response.Body)

        txtConsole.SelectionColor = Color.White
        txtConsole.AppendText(Microsoft.VisualBasic.Constants.vbCrLf & Microsoft.VisualBasic.Constants.vbCrLf & "> ")
        txtConsole.ScrollToCaret()
    End Sub


    Private Sub btnSendCommand_Click(sender As Object, e As EventArgs)
        Dim response = _ur.InterpreterMode.ExecuteCommand(txtCommand.Text)
        Log(response)
    End Sub

    Private Sub btnEndInterpreter_Click(sender As Object, e As EventArgs)
        Dim response = _ur.InterpreterMode.EndInterpreter()
        Log(response)
    End Sub

    Private Sub btnClearInterpreter_Click(sender As Object, e As EventArgs)
        Dim response = _ur.InterpreterMode.ClearInterpreter()
        Log(response)
    End Sub

    Private Sub btnAbort_Click(sender As Object, e As EventArgs)
        Dim response = _ur.InterpreterMode.Abort()
        Log(response)
    End Sub

    Private Sub btnSkipBuffer_Click(sender As Object, e As EventArgs)
        Dim response = _ur.InterpreterMode.SkipBuffer()
        Log(response)
    End Sub

    Private Sub btnLastExecuted_Click(sender As Object, e As EventArgs)
        Dim response = _ur.InterpreterMode.StateLastExecuted()
        Log(response)
    End Sub

    Private Sub btnLastInterpreted_Click(sender As Object, e As EventArgs)
        Dim response = _ur.InterpreterMode.StateLastInterpreted()
        Log(response)
    End Sub

    Private Sub btnLastCleared_Click(sender As Object, e As EventArgs)
        Dim response = _ur.InterpreterMode.StateLastCleared()
        Log(response)
    End Sub

    Private Sub btnLastUnexecuted_Click(sender As Object, e As EventArgs)
        Dim response = _ur.InterpreterMode.StateLastUnexecuted()
        Log(response)
    End Sub
End Class
